File: Algorithms/Rust/src/megnav/algorithms/62_unique_paths.rs
// LeetCode 62. Unique Paths
struct Solution;

impl Solution {
    pub fn unique_paths(m: i32, n: i32) -> i32 {
        let mut array = vec![vec![-1; n as usize]; m as usize];
        Solution::recursive(m, n, 0, 0, &mut array)
    }

    pub fn recursive(m:i32, n:i32, i:i32, j:i32, array:&mut Vec<Vec<i32>>) -> i32 {
        if i == m-1 && j == n-1 {
            return 1;
        }
        if i >= m || j >= n {
            return 0;
        }

        if array[i as usize][j as usize] != -1 {
            return array[i as usize][j as usize];
        }

        array[i as usize][j as usize] = Solution::recursive(m,n, i+1, j, array) + Solution::recursive(m,n, i, j+1, array);
        array[i as usize][j as usize]

    }
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_62() {
        let m = 3;
        let n = 7;
        let res = Solution::unique_paths(m, n);
        assert_eq!(res, 28);
    }
}

fn main() {}

File: Language/Python/LeetCode/62_unique_paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        # (m+n-2)C(m-1)
        num = n+m-2
        r = m-1
        res = 1
        for i in range(1, r+1):
            res = res*(num-r+i)/i
        return int(res)


class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        row = [0] * (n+1)
        row[n-1] = 1
        for i in range(m-1, -1, -1):
            for j in range(n-1, -1, -1):
                row[j] = row[j] + row[j+1]

        return row[0]
